File: src/Chapter04GraphBFS/GraphBFS.java
package Chapter04GraphBFS;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

// Time Complexity: O(V+E)
public class GraphBFS {

    private Graph g;
    private boolean[] visited;

    private ArrayList<Integer> order = new ArrayList<>();

    public GraphBFS(Graph g) {
        this.g = g;
        visited = new boolean[g.V()];
        for (int v = 0; v < g.V(); v++) {
            if (!visited[v])
                bfs(v);
        }
    }

    public void bfs(int s) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(s);
        visited[s] = true;
        while (!queue.isEmpty()) {
            int v = queue.remove();
            order.add(v);

            for (int w : g.adj(v))
                if (!visited[w]) {
                    queue.add(w);
                    visited[w] = true;
                }
        }
    }

    public Iterable<Integer> order() {
        return order;
    }

    public static void main(String[] args) {

        Graph g = new Graph("/Users/hui/Desktop/java/play-with-graph-algorithme/src/Chapter04GraphBFS/g.txt");
        GraphBFS graphBFS = new GraphBFS(g);
        System.out.println("BFS Order : " + graphBFS.order());
    }
}

File: src/Chapter04GraphBFS/python/GraphBFS.py
from Chapter02GraphDFS.python.adjSet import adjSet as Graph
from collections import deque


class GraphBFS:
    def __init__(self, G):
        self._G = G
        self._visited = [False] * G.V
        self._order = []

        for v in range(G.V):
            if not self._visited[v]:
                self._bfs(v)

    def _bfs(self, s):
        queue = deque()
        queue.append(s)
        self._visited[s] = True

        while queue:
            v = queue.popleft()
            self._order.append(v)
            for w in self._G.adj(v):
                if not self._visited[w]:
                    queue.append(w)
                    self._visited[w] = True

    def order(self):
        return self._order


if __name__ == "__main__":
    filename = "/Chapter03GraphDFSApplications/g.txt"
    g = Graph(filename)

    path = GraphBFS(g)
    print(path.order())
